config/src/main/java/org/springframework/security/config/annotation/web/configuration/Saml2ServiceProviderConfiguration.java

```diff
@@ -0,0 +1,152 @@
+import org.joda.time.DateTime;
+import org.opensaml.core.config.ConfigurationService;
+import org.opensaml.core.config.InitializationException;
+import org.opensaml.core.config.InitializationService;
+import org.opensaml.core.xml.XMLObjectBuilderFactory;
+import org.opensaml.core.xml.config.XMLObjectProviderRegistry;
+import org.opensaml.saml.common.xml.SAMLConstants;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.impl.AuthnRequestBuilder;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.context.annotation.ImportSelector;
+import org.springframework.core.Ordered;
+import org.springframework.core.type.AnnotationMetadata;
+import org.springframework.security.saml2.Saml2Exception;
+import org.springframework.util.ClassUtils;
+
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+import static org.opensaml.core.xml.config.XMLObjectProviderRegistrySupport.getBuilderFactory;
+import static org.opensaml.core.xml.config.XMLObjectProviderRegistrySupport.setParserPool;
+
+/**
+ * {@link Configuration} for SAML 2.0 Service Provider support
+ *
+ * <p>
+ * This {@code Configuration} is conditionally imported by {@link Saml2ImportSelector}
+ * when the {@code spring-security-saml2-service-provider} module is present on the classpath.
+ *
+ * @author Josh Cummings
+ * @since 5.4
+ * @see Saml2ImportSelector
+ */
+@Import(Saml2ServiceProviderConfiguration.OpenSamlImportSelector.class)
+final class Saml2ServiceProviderConfiguration {
+
+	static class OpenSamlImportSelector implements ImportSelector {
+
+		@Override
+		public String[] selectImports(AnnotationMetadata importingClassMetadata) {
+			boolean opensamlPresent = ClassUtils.isPresent(
+					"org.opensaml.core.config.InitializationService", getClass().getClassLoader());
+
+			return opensamlPresent ?
+					new String[] { "org.springframework.security.config.annotation.web.configuration.Saml2ServiceProviderConfiguration.OpenSamlServiceProviderConfiguration" } :
+					new String[0];
+		}
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	static class OpenSamlServiceProviderConfiguration {
+		@Bean
+		BeanFactoryPostProcessor saml2BeanFactoryPostProcessor() {
+			return new OpenSamlServiceProviderBeanFactoryPostProcessor();
+		}
+
+		private static class OpenSamlServiceProviderBeanFactoryPostProcessor
+				implements BeanFactoryPostProcessor, Ordered {
+
+			@Override
+			public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {
+				if (ConfigurationService.get(XMLObjectProviderRegistry.class) == null) {
```

Comment:
@rwinch What are your thoughts about initializing OpenSAML in `spring-security-config`? What I'm trying to achieve is remove it from the static initialization of `OpenSamlImplementation`, and this seems to be the easiest way to hide the implementation while still giving the application a hook to override what's done here.

As a side note, I think there's an argument to be made for Spring Security not initializing OpenSAML at all; however, 5.2. and 5.3 both initialize OpenSAML, so it would be a surprise to upgrade to 5.4 and find that no longer happening.

Reply:
Also, of course, feel free to comment on other parts of the PR, but the above is my main question. If you agree that this is a viable route, then it's the one I'm planning on taking.

Reply:
I really don't like pushing this into configuration. It makes it harder to use without Spring configuration and also complicates things a bit (now users need to understand BeanFactoryPostProcessor to understand the initialization). Furthermore, I don't like the initialization tied to configuration. If someone creates a new `OpenSamlAuthenticationProvider` it should work without being tied to the Bean container.

Instead, I'd prefer exposing a single static method that initializes Open SAML that any class leveraging Open SAML could invoke. 

I think ideally we'd remove the static initialization all together. Perhaps SAML has a way to provide the `ParserPool` as an argument vs using a default static value?